webapp-next/webserver/webserver.go

```diff
@@ -0,0 +1,137 @@
+package webserver
+
+import (
+	"app/calc"
+	"app/utils"
+	"bytes"
+	"github.com/gin-gonic/gin"
+	ginprometheus "github.com/zsais/go-gin-prometheus"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+)
+
+type ReportError struct {
+	Code  int
+	Inner error
+}
+
+func (r ReportError) Error() string {
+	return r.Inner.Error()
+}
+
+type ApiError struct {
+	Error string `json:"detail"`
+}
+
+type bodyLogWriter struct {
+	gin.ResponseWriter
+	body *bytes.Buffer
+}
+
+func (w bodyLogWriter) Write(b []byte) (int, error) {
+	w.body.Write(b)
+	return w.ResponseWriter.Write(b)
+}
+
+func ProxyCompareMw(c *gin.Context) {
+	var proxyBuf bytes.Buffer
+	var logBuf bytes.Buffer
+
+	tmpTee := io.TeeReader(c.Request.Body, &proxyBuf)
+	tee2 := io.TeeReader(tmpTee, &logBuf)
+
+	// Actual request processing
+	c.Request.Body = ioutil.NopCloser(tee2)
+	blw := &bodyLogWriter{body: bytes.NewBufferString(""), ResponseWriter: c.Writer}
+	c.Writer = blw
+	c.Next()
+
+	// Proxy request processing
+	destUrl := utils.MustGetEnv("PROXY_ADDRESS") + c.Request.URL.Path
+	utils.Log("url", destUrl).Info("Calling")
+
+	proxyReq, err := http.NewRequest(c.Request.Method, destUrl, &proxyBuf)
+	if err != nil {
+		panic(err)
+	}
+	proxyReq.Header = c.Request.Header
+	resp, err := http.DefaultClient.Do(proxyReq)
+	if err != nil {
+		panic(err)
+	}
+	respData, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		panic(err)
+	}
+
+	// Logging for analysis, skip metrics
+	if !strings.HasSuffix(c.Request.URL.Path, "metrics") {
+		utils.Log("path", c.Request.URL.Path, "go", blw.body.String(), "py", string(respData), "req", logBuf.String()).Info("Call")
+	}
+}
+
+func Proxy(c *gin.Context) {
+	utils.Log().Info("Calling proxy")
+	remote, err := url.Parse(utils.MustGetEnv("PROXY_ADDRESS"))
+	if err != nil {
+		panic(err)
+	}
+	proxy := httputil.NewSingleHostReverseProxy(remote)
+	c.Request.URL.Host = remote.Host
+	c.Request.Host = remote.Host
+	proxy.ServeHTTP(c.Writer, c.Request)
+}
+
+func Run() {
+	r := gin.Default()
+	r.Use(gin.Logger())
+	r.RedirectTrailingSlash = true
+	r.Use(ProxyCompareMw)
+	r.Use(gin.ErrorLogger())
+	r.GET("/api/v1/version", func(c *gin.Context) {
+		c.String(200, "1.12.2%v", "")
```

Comment:
The version needs to be somehow integrated with semantic-release-python. :smile: